### PR TITLE
(#104, #16) Trial type redirects, Disease & Trial Type display

### LINKED ISSUE #16: Story: Display Listing Results for Disease & Trial Type
## As a user, I would like to view the trial results for valid disease and trial type code parameters on a dynamic trial listing page, along with links to the trial's description page, the brief summary of the trial, and the number of locations of the trial.

This should include:
- handling the routing for the disease dynamic URL
- fetching the ListingInformation for the given disease parameter
- fetching the TrialTypeInfo for the given trial type parameter
- fetching the results from the CTS API
- displaying the title, intro text, and listing results
- pager
- setting the metadata
- analytics
- page not found when disease or trial type does not exist
- No Trials Found redirect page
- Redirect ugly URLs to pretty URLs

The title and intro text for the disease and trial type listing page is set according to given parameters.
H1: [Trial Type] Clinical Trials for [Disease Name]
Intro text:
```
<p>Clinical trials are research studies that involve people. The clinical trials on this list are for [disease name] [trial type]. All trials on the list are NCI-supported clinical trials, which are sponsored or otherwise financially supported by NCI.</p><p>NCI’s <a href=\"/about-cancer/treatment/clinical-trials/what-are-trials\">basic information about clinical trials</a> explains the types and phases of trials and how they are carried out. Clinical trials look at new ways to prevent, detect, or treat disease. You may want to think about taking part in a clinical trial. Talk to your doctor for help in deciding if one is right for you.</p>
```

The metadata for the disease and trial type listing pages is set according to the given parameters.
Title: [Trial Type] Clinical Trials for [Disease Name] - National Cancer Institute
Og title: [Trial Type] Clinical Trials for [Disease Name] Clinical Trials
Og url: [base host]/[disease-friendly-name-or-codes]/[trial-type]
Description: "NCI supports clinical trials studying new and more effective ways to detect and treat cancer. Find [trial type] clinical trials for [disease name]."
Canonical URL: [canonical-host]/[disease-friendly-name-or-codes]/[trial type]

Steal the locations code from the CTS App.

> **ESTIMATE** 3

#### General page
```gherkin
Scenario: View dynamic listing page results with disease and trial type parameters
  Given the user navigates to "/C4872/treatment"
  And "trialListingPageType" is set to "Disease"
  Then the page displays the title 'Treatment Clinical Trials for Breast Cancer'
    And the page displays the intro text under the title
    And  the page displays 'Trials 1-25 of 916' under the intro text
    And each result displays the trial object's brief_title as a link to the trial's description page
    And each result displays the trial object's brief_summary below the link for the trial
    And each result displays 'Locations: # locations' below the description
```

#### Title/Intro Text
```gherkin
Scenario: View dynamic listing page title and intro text with disease and trial type parameters
Given the user navigates to "/C4872/treatment"
And "trialListingPageType" is set to "Disease"
  Then the system displays the page heading aka H1 as "Treatment Clinical Trials for Breast Cancer"
  Then the system displays the following introductory text below the H1: "Clinical trials are research studies that involve people. The clinical trials on this list are for breast cancer treatment. All trials on the list are NCI-supported clinical trials, which are sponsored or otherwise financially supported by NCI."
  Then the system displays a paragraph break
  Then the system displays a second paragraph: "NCI’s basic information about clinical trials explains the types and phases of trials and how they are carried out. Clinical trials look at new ways to prevent, detect, or treat disease. You may want to think about taking part in a clinical trial. Talk to your doctor for help in deciding if one is right for you."
    And the text "basic information about clinical trials" links to https://www.cancer.gov/about-cancer/treatment/clinical-trials/what-are-trials
  And the number of trials and pager displays below the intro text
  And the results appear below the number of trials and pager
```

#### Pager
``` gherkin
Scenario: Pager on dynamic listing page results with intervention and trial type parameters
  Given the user navigates to "/C1647/treatment"
  And "trialListingPageType" is set to "Intervention"
    Then the page title is "Treatment Clinical Trials Using Trastuzumab"
   Then the system displays "Trials 1-25 of" "57"
    And the top pager is aligned to left on the same line as the number of trials on desktop and tablet
   And the bottom pager is aligned to left 
    And pager displays the following navigation options
      | pages  |
      | 1      |
      | 2      |
      | 3      |
      | Next > |
    And the page "1" is highlighted
```
```gherkin
Scenario: Mobile Pager Display
  Given user navigates to a listing page results URL
  Then the system returns the results page for cancer
    Then a horizontal rule is displayed below the "Trials 1-25 of XXX" 
    Then a top pager is displayed aligned to the right
    Then a horizontal rule is displayed below the pager
    Then the result items are displayed 
    Then a horizontal rule is displayed 
    Then the bottom pager is displayed aligned to the right
```

#### Metadata
```gherkin
Scenario: View dynamic listing page metadata with disease and trial type parameters
  Given the site name is "National Cancer Institute"
     And "trialListingPageType" is set to "Disease"
     And the canonical host is "https://www.cancer.gov"
     And the base host is "http://localhost:3000"
When user navigates to "/C4872/treatment"
  Then the title tag should be "Treatment Clinical Trials for Breast Cancer - National Cancer Institute 
    And there is a meta tag with a property "og:title" and content "Treatment Clinical Trials for Breast Cancer"
    And there is a meta tag with a property "og:url" and content "http://localhost:3000/breast-cancer/treatment"
    And there is a meta tag with a property "description" and content "NCI supports clinical trials studying new and more effective ways to detect and treat cancer. Find treatment clinical trials for breast cancer."
    And there is a link tag with rel "canonical" and href  "https://www.cancer.gov/breast-cancer/treatment"
```

#### Analytics
```gherkin
  Scenario: Page Load Analytics fires when a user views a dynamic listing page with disease and trial type parameters
    Given "baseHost" is set to "http://localhost:3000"
       And "canonicalHost" is set to "https://www.cancer.gov"
       And "siteName" is set to "National Cancer Institute"
       And "channel" is set to "Clinical Trials"
       And "trialListingPageType" is set to "Disease"
    When the user navigates to "/C4872/treatment"
    Then there should be an analytics event with the following details
      | key                                          | value                                                                          |
      | type                                         | PageLoad                                                                       |
      | event                                        | TrialListingApp:Load:Results                                                   |
      | page.name                                    | www.cancer.gov/breast-cancer/treatment                                         |
      | page.title                                   | Treatment Clinical Trials for Breast Cancer                                    |
      | page.metaTitle                               | Treatment Clinical Trials for Breast Cancer - National Cancer Institute        |
      | page.language                                | english                                                                        |
      | page.type                                    | nciAppModulePage                                                               |
      | page.channel                                 | Clinical Trials                                                                |
      | page.contentGroup                            | Clinical Trials: Custom                                                        |
      | page.publishedDate                           | 02/02/2011                                                                     |
      | page.additionalDetails.numberResults         | (int) results total                                                        |
      | page.additionalDetails.trialListingPageType  | disease                                                                        |
      | page.additionalDetails.diseaseName           | breast cancer                                                                  |
      | page.additionalDetails.trialType             | treatment                                                                      |
      | page.additionalDetails.interventionName             | none                                                                      |

Scenario: Click event fires when a user clicks on result item
    Given "trialListingPageType" is set to "Disease"
    And "pageTitle" is set to "{{trial_type_label}} Clinical Trials for {{disease_label}}"
    And "analyticsPublishedDate" is set to "02/02/2011"
    When the user navigates to "/breast-cancer/treatment"
    Then the page title is "Treatment Clinical Trials for Breast Cancer"
    When user clicks on result item 1
    Then there should be an analytics event with the following details
      | key              | value                             |
      | type             | Other                             |
      | event            | TrialListingApp:Other:ResultClick |
      | linkName         | CTSLink                           |
      | data.resultIndex | (int)1                            |
      | data.currentPage | (int)1                            |

```

#### Page Not Found for invalid Disease or Trial Type
```gherkin
Scenario: Page Not Found displays when a user views a dynamic listing page with invalid disease or trial type parameters
Given user navigates to “/C123455/chicken”
And "trialListingPageType" is set to "Disease"
When the user has entered an incorrect disease or trial type
Then the system redirects users to a 404 page 
     Then page title on error page is "Page Not Found"
     And the search bar appears below the text
     And the following links and texts exist on the page: "We can't find the page you're looking for." Visit the homepage, browse by cancer type, or use the search below." "Have a question? Get in touch."
  
      | https://www.cancer.gov          | homepage      | 
      | https://www.cancer.gov/types    | cancer type   | 
      | https://www.cancer.gov/contact  | Get in touch  |

     And the system returns a 404 status code
```
```gherkin
Given the application returns a "Page not found" message
  Then the page should contain the markup <meta name="prerender-status-code" content="404">
```

```gherkin
Scenario: Page Load Analytics fires for a 404 with invalid disease or trial type
 Given "trialListingPageType" is set to "Disease"
    And "baseHost" is set to "http://localhost:3000"
    And "canonicalHost" is set to "https://www.cancer.gov"
    And "analyticsPublishedDate" is set to "02/02/2011"
    And "siteName" is set to "National Cancer Institute"
 When the user navigates to "/C123456/chicken"
   Then page title on error page is "Page Not Found"
    And browser waits
   Then there should be an analytics event with the following details

| key                                          | value                                            |
| type                                         | PageLoad                                         |
| event                                        | TrialListingPageApp:Load:PageNotFound            |
| page.name                                    | www.cancer.gov/C123456/chicken                   |
| page.title                                   | Page Not Found                                   |
| page.metaTitle                               | Page Not Found                                   |
| page.language                                | english                                          |
| page.type                                    | nciAppModulePage                                 |
| page.channel                                 | Clinical Trials                                  |
| page.contentGroup                            | Clinical Trials: Custom                          |
| page.additionalDetails.trialListingPageType  | disease                                          |



```

#### No Trials Found page
```gherkin
Scenario: Page redirects and displays No Trials Found Page for Disease and Trial Type
Given user navigates to “/C1234/treatment”
And "trialListingPageType" is set to "Disease"
  When no trials are returned for the disease and trial type
  Then the system redirects users to /notrials?p1=spiroplatin&p2=treatment
  Then the system displays the H1
  Then the system displays the message: “There are no NCI-supported clinical trials for spiroplatin treatment at this time. You can try a new search or contact our Cancer Information Service to talk about options for clinical trials.”
Then the system displays the banner for “Questions about Clinical Trials?” 
```
```gherkin
Scenario: Status Code Returned for no trials found for Disease and Trial Type
Given user navigates to “/C1234/treatment”
And "trialListingPageType" is set to "Disease"
When no trials are returned for the disease and trial type
Then the system redirects users to /notrials?p1=spiroplatin&p2=treatment
Then the page should return a 404 status code.
```

#### Redirect ugly URL to pretty URL
```gherkin
Scenario: Page redirects to pretty URL if codes are given for Disease and Trial Type
Given user navigates to “/C4872/treatment”
And "trialListingPageType" is set to "Disease"
  When a mapping exists for C4872
  Then the system should redirect the URL to “/breast-cancer/treatment”
```

### Resources:
* Feature: JIRA Ticket #
* Mockups: *Here goes a URL to or the name of the mockup(s) in inVision*
* Web Page: *Here goes a URL to an existing web page*

### Notes

*Some complementary notes if necessary:*

* > Here goes a quote from an email
* Here goes whatever useful information can exist…

---

### Prerequisites
- *required dependency for identifying the problem, scope and then solution for this issue*
- [ ] Short Spike Description - Issue #9999

### Tasks
- [ ] Add `CodeOrPurlWithTypePath` route to `routing.js` for Disease & Trial Type
```js
const appPaths = {
	...,
	CodeOrPurlWithTypePath: '/:codeOrPurl/:type',
	...
};
```
- [ ] Update  disease component 
- [ ] Update unit test
- [ ] Add route params for disease trial type
- [ ] Update no trials found
- [ ] Add integration tests

### Technical Notes
- More notes, or technical nuances/direction

### LINKED ISSUE #104: Trial type parameters given as idString should redirect to prettyUrlName
### Issue description
Currently, if you go to a Disease or Intervention dynamic listing with the trial type parameter given as the idString, it does not redirect you to the prettyUrlName.
Example: /breast-cancer/supportive_care/trastuzumab?cfg=0

And URLs with the idString given as the trial type parameter should redirect to the prettyUrlName for that trial type.
supportive_care -> supportive-care
health_sciences_research -> health-sciences-research
basic_science -> basic-science

> **ESTIMATE** TBD

### What's the expected change?
-

### What's the current functionality?
-

### What's the updated acceptance criteria?
```gherkin
```

### Additional details / screenshot
- Screenshot

### Related Tickets
* Original Story Ticket #9999
* Original Story Ticket #9999

### PR DESCRIPTION
**Pull request by sarinapadilla**
(#16) Disease and Trial Type Display

* Add CodeOrPurlWithTypePath route
* Add route to App.js
* * Update data-route-param-util to set correct value for trial-type param
* Add unit tests for DiseaseTrialType
* Add integration tests for DiseaseTrialType
* Add integration test data for DiseaseTrialType
* Update integration tests to account for new params structure
* Closes #16

(#104) Trial type parameter redirects

* Update getTrialType to allow underscores for trial type requests
* Update getTrialType unit tests
* Update CTLViewsHoC to redirect trial type idStrings to prettyUrlNames
* Update CTLViewsHoC unit tests with trial type redirects
* Closes #104

Notes:
* If you visit a URL with idString for a trial type and are routed to the Page Not Found, the redirect from idString -> prettyUrlName does not happen (like current Prod functionality).
* If you visit a URL with idString for a trial type and are routed to the No Trials Page, the prettyUrlName should be the p2 parameter for trial type in the No Trials URL.

### CODE REVIEW

**Comment by sarinapadilla** [comment]
@VictoriaSunNIH @alinai @mworrest this PR is ready for testing: https://react-app-dev.cancer.gov/clinical-trials-listing-app/pr-106/breast-cancer/supportive-care?cfg=0

Please also check that trial type redirects are working. Examples are:
* supportive_care to supportive-care
* health_sciences_research to health-sciences-research
* basic_science to basic-science
These redirects should append the redirect=true parameter to the URL.

**Comment by VictoriaSunNIH** [comment]
@sarinapadilla , @mworrest  - I've verified that these are working beautifully! Thank you!

File: src/views/CTLViewsHoC.jsx [hunk 1]
```diff
@@ -125,9 +125,13 @@ const CTLViewsHoC = (WrappedView) => {
 				// handle redirects if we are not showing /notrials, for
 				// sanity.
 				if (!isNoTrials) {
+					console.log(getListingInfo.payload);
```
Comment (kate-mashkina) [inline src/views/CTLViewsHoC.jsx:128]: could you remove this console log?

File: src/views/CTLViewsHoC.jsx [hunk 2]
```diff
@@ -151,6 +155,8 @@ const CTLViewsHoC = (WrappedView) => {
 										};
 									}
 									case 'trial-type': {
+										// const ttPayload = getListingInfo.payload[idx];
```
Comment (kate-mashkina) [inline src/views/CTLViewsHoC.jsx:158]: same here  - probably you don't need this commented out code

**Review by kate-mashkina — COMMENTED** [review]
just the console.log and commented out code and its good to go!

**Comment by alinai** [comment]
Looks good to me. Thanks @sarinapadilla!

File: src/views/CTLViewsHoC.jsx (hunk 1 above)
Comment (sarinapadilla) [reply src/views/CTLViewsHoC.jsx]: Nice catch - removed!

File: src/views/CTLViewsHoC.jsx (hunk 2 above)
Comment (sarinapadilla) [reply src/views/CTLViewsHoC.jsx]: Also removed!

**Review by seyilonge-nci — APPROVED** [review]

**Review by kate-mashkina — APPROVED** [review]

**Pull request merged**